### ISSUE TITLE
TF 1.12: nvcc fails: absl/strings/string_view.h(501): error: constexpr function return is non-constant

### ISSUE THREAD

**Issue by abhigarg-iitk** [rwth-i6/returnn#87]
I am having issues with running returnn experiments with TF1.12 and Python 2. My error is this

`CalledProcessError: Command '['/ml_env/cuda9/cuda-9.0/bin/nvcc', '-shared', '-O2', '-std=c++11', '-I',  '/vol0/ml_env/cuda9/.venv/py2/asr/lib/python2.7/site-packages/tensorflow/include', '-I', '/vol0/ml_env/cuda9/.venv/py2/asr/lib/python2.7/site-packages/tensorflow/include/external/nsync/public', '-I', '/t ool/ml/cuda9/cuda-9.0/include', '-L', '/ml_env/cuda9/cuda-9.0/lib64', '-x', 'cu', '-DGOOGLE_CUDA=1', ' -Xcompiler', '-fPIC', '-arch', 'compute_61', '-D_GLIBCXX_USE_CXX11_ABI=0', '-g', '/tmp/abhinav.garg/ret urnn_tf_cache/ops/NativeLstm2/734b69db29/NativeLstm2.cc', '-o', '/tmp/abhinav.garg/returnn_tf_cache/ops /NativeLstm2/734b69db29/NativeLstm2.so', '-L/vol0/ml_env/cuda9/.venv/py2/asr/lib/python2.7/site-packag es/numpy/.libs', '-lopenblasp-r0-39a31c03.2.18', '-L/vol0/ml_env/cuda9/.venv/py2/asr/lib/python2.7/sit e-packages/tensorflow', '-ltensorflow_framework']' returned non-zero exit status 1 
`

Which is very similar to https://github.com/tensorflow/tensorflow/issues/22766. I found the solution of using `-DNDEBUG` but I'm not sure if it's the correct solution. Can you please tell me how can I run returnn with tf1.12. I have tried applying this patch https://github.com/rwth-i6/returnn/commit/5913f1f975e08f1172217e75dbe5aa4263068a84 but it's still giving me the same error.

Thanks in Advance

**Comment by patrick-wilken** [comment]
For the record, I'm also using `-DNDEBUG` at the moment as a workaround:

```
  def get_compiler_opts(self):
    return [
      "-I", "%s/include" % self.cuda_path,
      "-L", "%s/%s" % (self.cuda_path, self._get_lib_dir_name()),
      "-x", "cu",
      "-v",
      "-DNDEBUG"]
```
in` TFUtil.py`.

**Comment by abhigarg-iitk** [comment]
@patrick-wilken  Even after using this flag I'm unable to run the training. I'm getting the following error

`
 InternalError: cudnn PoolBackward launch failed
      [[{{node optimize/gradients/slstm2_pool/max_pool_grad/MaxPoolGrad}} = MaxPoolGrad[T=DT_FLOAT, data_format="NHWC", ksize=[1, 1, 2, 1], padding="SAME", strides=[1, 1, 2, 1], _device="/job:localhost/replica:0
 /task:0/device:GPU:0"](slstm2_pool/max_pool/ExpandDims, slstm2_pool/max_pool, optimize/gradients/slstm2_pool/max_pool/Squeeze_grad/Reshape)]]
`

This mainly occurs during backward pass. Forward pass (dev-eval) works fine but during training it throws this error.

Did you also encounter such error ?

**Comment by albertz** [comment]
@abhigarg-iitk Which TensorFlow version is that? Also, what CUDA/cuDNN version? And what GPU?

Related might be https://github.com/tensorflow/tensorflow/issues/13292. Or maybe https://github.com/tensorflow/tensorflow/issues/6698. When you search a bit, you find some other people reporting this.

**Comment by abhigarg-iitk** [comment]
@albertz Indeed it was an CUDA/cuDNN version mismatch error.

Earlier I was using : TF1.12, CUDA9, cudnnV7
Now I'm using : TF1.12, CUDA9.2, cuDNNv7.2

However kindly let me know if using -DNDEBUG is the correct solution?

**Comment by albertz** [comment]
`-DNDEBUG` is a workaround (which works, but is ugly, and it will disable some internal checks).
The problem has to be fixed by TensorFlow (not in Returnn). It's exactly the bug you linked: https://github.com/tensorflow/tensorflow/issues/22766 (which I reported)

**Comment by albertz** [comment]
Better leave this open for now, until this is really fixed (either on TF side, or on our side).

**Comment by albertz** [comment]
@curufinwe I think you said that Google officially states that for compiling custom C++ ops, you should always use `-DNDEBUG`? Is that right? Do you have a ref?

**Comment by albertz** [comment]
Btw, also see [this TF bug](https://github.com/tensorflow/tensorflow/issues/17316), which is also triggered by **not** using `-DNDEBUG`. This is why we use it in `TFOpenFst.py` and `TFKenLM.py`.

**Comment by albertz** [comment]
Fixed now via 9d17c82619c07fe.
